File: 2019-11-25/JIIIIIND/구명보트.java
import java.util.Collections;
import java.util.ArrayList;


class Solution {
    public int solution(int[] people, int limit)
    {
        int answer = 0;
        ArrayList<Integer> boatPeople = new ArrayList<>();
        for(int value : people)
            boatPeople.add(value);
        //오름차순 정렬 됨
        Collections.sort(boatPeople);

        while(!boatPeople.isEmpty())
        {
            int first = boatPeople.get(boatPeople.size()-1);
            int nextIndex = -1;
            for(int i = 0; i < boatPeople.size()-1; i++)
            {
                if(boatPeople.get(i) + first <= limit)
                    nextIndex = i;
                else
                {
                    if(nextIndex >= 0)
                    {
                        boatPeople.remove(nextIndex);
                        nextIndex = -1;
                    }
                    break;
                }
            }
            if(nextIndex >= 0)
            {
                boatPeople.remove(nextIndex);
                boatPeople.remove(boatPeople.size()-1);
                answer++;
            }
            else
            {
                boatPeople.remove(boatPeople.size()-1);
                answer++;
            }
        }
        return answer;
    }
}

File: 2019-11-25/구명보트.cpp
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

int solution(vector<int> people, int limit)
{
	int answer = 0;
	int end = people.size() - 1;
	vector<bool> is_checked(people.size(), false);

	// 오름차순으로 정렬하고, 제일 몸무게가 낮은 사람부터 골라 최대한 몸무게를 많이 채울 수 있는 방향으로 계산한다.
	std::sort(&people[0], &people[0] + people.size());

	for (int i = 0; i people.size(); i++)
	{
		int pass_light = people[i];

		if (is_checked[i])
			continue;

		for (int j = end; j i; j--)
		{
			if (!is_checked[j] && pass_light + people[j] = limit)
			{
				end = j - 1;
				is_checked[i] = true;
				is_checked[j] = true;
				answer++;
				break;
			}
		}
	}

	for (int i = 0; i people.size(); i++)
	{
		if (!is_checked[i])
		{
			is_checked[i] = true;
			answer++;
		}
	}

	return answer;
}